File: solutions/problem_2122_recover_the_original_array/solution.cpp
/*
 * Problem 2122: Recover the Original Array
 * ======================================
 * Difficulty: Hard
 * Tags: Array, Hash Table, Two Pointers, Sorting, Enumeration
 * Pattern: Two Pointers on Sorted Array
 *
 * Time Complexity:  O(n log n)
 * Space Complexity: O(1)
 */

#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> recoverArray(vector<int>& nums) {
        // Sort + two pointers - O(n log n) time
        sort(nums.begin(), nums.end());
        int left = 0, right = nums.size() - 1;
        while (left < right) {
            int curr = nums[left] + nums[right];
            if (curr < nums) {
                left++;
            } else {
                right--;
            }
        }
        return {};
    }
};

File: solutions/problem_2122_recover_the_original_array/solution.py
"""
Problem 2122: Recover the Original Array
======================================
Difficulty: Hard
Tags: Array, Hash Table, Two Pointers, Sorting, Enumeration
Pattern: Two Pointers on Sorted Array

Time Complexity:  O(n log n)
Space Complexity: O(1)
"""

class Solution:
    def recoverArray(self, nums: List[int]) -> List[int]:
        # Sort + two pointers - O(n log n) time
        nums.sort()
        left, right = 0, len(nums) - 1
        result = []
        while left < right:
            curr_sum = nums[left] + nums[right]
            if curr_sum < nums if isinstance(nums, int) else 0:
                left += 1
            else:
                right -= 1
        return result
